src/extra/components/dld.py

```diff
@@ -0,0 +1,389 @@
+            # Missing counts in one or more trains in data, select data
+            # and pulses down.
+            from extra_data import by_id
+            train_sel = by_id[entry_counts.index.to_numpy()]
+
+            pulses = self._pulses.select_trains(train_sel)
+            kd = kd.select_trains(train_sel)
+        else:
+            pulses = self._pulses
+
+        return pulses.build_pulse_index(pulse_dim)
+
+    def _build_reduced_pd(self, data, index, entry_level):
+        """Get variable-length data as series or dataframe.
+
+        Load and reduce data with a variable number of actual entries
+        among a fixed number of potential entries to build a linear
+        pandas Series or DataFrame.
+
+        The data must be 2-dimensional, with the first axis is assumed
+        to be an entry ("pulse") dimension and the second to only
+        contain actual data up to a variable number for each entry.
+
+        The data type must either be a scalar floating type (returning
+        Series objects) or a structured data type (returning DataFrame
+        objects) containing at least a single floating type.
+
+        Args:
+            key (str): Instrument source key.
+            pulse_dim (str): Pulse to use in index, must be supported by
+                underlying pulses object.
+            entry_level (str or None): Additional index level inserted
+                for each entry or omitted if None.
+
+        Returns:
+            (pandas.Series or pandas.DataFrame): Series objects are
+                returned for scalar data and DataFrame objects for
+                structured data.
+        """
+
+        num_rows = data.shape[1]
+        raw = data.ravel()
+
+        # Obtain mask which entries are actually filled.
+        if data.dtype.fields is not None:
+            pd_cls = pd.DataFrame
+            field = {name for name, (dtype, offset)
+                    in data.dtype.fields.items()
+                    if np.issubdtype(dtype, np.floating)}.pop()
+
+            finite_mask = np.isfinite(raw[field])
+        elif np.issubdtype(data.dtype, np.floating):
+            pd_cls = pd.Series
+            finite_mask = np.isfinite(raw)
+        else:
+            raise TypeError(data.dtype)
+
+        if data.size == 0 or index is None:
+            return pd_cls(np.zeros(0, dtype=data.dtype))
+
+        # Obtain pulse index and repeat each row by the number of
+        # actual data in each entry, then convert to a dataframe.
+        index_df = index.repeat(
+            finite_mask.reshape(-1, num_rows).sum(axis=1)).to_frame()
+
+        # Insert additional index level enumerating actual entries.
+        if entry_level is not None:
+            index_df[entry_level] = np.flatnonzero(finite_mask) % num_rows
+
+        return pd_cls(raw[finite_mask], pd.MultiIndex.from_frame(index_df))
+
+    def _get_reduced_pd(self, key, pulse_dim, entry_level, split_axis):
```

Comment:
I think this method is no longer called anywhere?

Reply:
Certainly, for comparison with the new version.